trezorlib/cardano.py

```diff
@@ -0,0 +1,130 @@
+# This file is part of the Trezor project.
+#
+# Copyright (C) 2012-2018 SatoshiLabs and contributors
+#
+# This library is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License version 3
+# as published by the Free Software Foundation.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the License along with this library.
+# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
+
+import base64
+import struct
+import binascii
+
+from . import messages
+from . import tools
+from .client import field
+from .client import expect
+from .client import session
+
+
+@field('address')
+@expect(messages.CardanoAddress)
+def get_address(client, address_n, show_display=False):
+    return client.call(
+        messages.CardanoGetAddress(
+            address_n=address_n, show_display=show_display))
+
+
+@expect(messages.CardanoPublicKey)
+def get_public_key(client, address_n):
+    return client.call(
+        messages.CardanoGetPublicKey(address_n=address_n)
+    )
+
+
+@expect(messages.CardanoMessageSignature)
+def sign_message(client, address_n, message):
+    return client.call(
+        messages.CardanoSignMessage(
+            address_n=address_n,
+            message=message.encode()
+        )
+    )
+
+
+def verify_message(client, public_key, signature, message):
+    return client.call(
+        messages.CardanoVerifyMessage(
+            public_key=binascii.unhexlify(public_key),
+            signature=binascii.unhexlify(signature),
+            message=message.encode()
+        )
+    )
+
+
+@session
+def cardano_sign_transaction(client, transaction):
+    msg = create_sign_transaction_message(transaction)
+
+    response = client.call(msg)
+    while response.tx_index is not None:
+        tx_index = response.tx_index
+
+        ack_message = create_tx_ack_message(transaction.get('transactions'), tx_index)
```

Comment:
the `create_tx_ack_message` is sort of pointless, esp. if you get `transactions` directly as an argument. please inline it here.

(also maybe verify that the requested index makes sense? but AFAICT the `index` is pointless anyway, so maybe it doesn't matter)

Reply:
fixed without index verification